Rewrite outgoing links diff to not use `html_token`

Mr0grog opened the pull request:
It turns out the `html_token` diff was not as great a basis for this as I’d thought and it generated diffs that were not as clear as they could be for analysts.

Instead, rewrite the diff to generate a list of `Link` objects, which can then be run through `difflib.SequenceMatcher` for diffing. Then we format the output of that (as a bonus, we can easily get optional JSON formatting now).

I also added some extra fancy footwork by making this a two-stage model: `Link` objects compare as equal even if only their text or only their `href` attributes match. When we hit an “equal” link that wasn’t an *exact* match, we do a second pass with diff-match-patch to diff the text and `href` inside it.

![linksdiff](https://user-images.githubusercontent.com/74178/35366078-70556a4a-012c-11e8-9feb-91b33d05b693.png)

Rows have a light red BG if they only contain removals, a light green BG if only additions, and gray if they contain both.

Fixes #151

ericnost commented:
I have some color deficiencies, making it hard for me to identify the color value of the less saturated BGs. I can distinguish a _difference_ in the two links in the bottom block of the example above, but I can't _name_ the color. I also can't specify what color the light BGs are in the first two blocks of the example. I am not sure how often I will run into this use case scenario in my monitoring.

I do better with the solid red/green BGs, especially when there are substantial chunks of text highlighted. (And I don't think these are pure red/green to begin with?) When there are small text changes right next to each other, I sometimes can't immediately tell whether it's a removal or addition. 

I don't feel burdened in my own monitoring work (which is why I haven't raised the issue before), but it might be worth testing an alternative, color-blind friendly color scheme for a broader audience. I am not sure what that would be, but there are some ideas here: http://www.visualisingdata.com/2015/11/colour-swatch-alternatives-to-green-and-red/ (maybe the first pair?)

weatherpattern commented:
@ericnost you raise a great and important point. Thanks for bringing it up and sharing the link, which is great. (How could one that opens with the colorbrewer not be?)

I like the first choice. # 4dac26 and # d01c8b
That way we don't have to reteach what green means, and the magenta color is not that far removed from the red that we're used to seeing.

Mr0grog commented on web_monitoring/links_diff.py:
```diff
@@ -1,8 +1,12 @@
 
-def links_diff(a_text, b_text):
+def links_diff(a_text, b_text, json=False):
     """
```
Well… it’s a kwarg because this is the HTTP interface, too. Should I separate out a library-focused entry-point to this module from the diff *server* focused entry point?

Mr0grog replied:
(Side note: would be great to get a standardized format param to these diffs… e.g. `/{diff name}.{format kwarg}` or maybe basing it off the `Accept` request header.)

Mr0grog replied:
I guess we could just register two endpoints with two funcs, one at `/links` → `links_diff_html()` and one at `/links.json` → `links_diff.links_diff()`. Do you think that’s preferable?

danielballan replied:
I like that better, yeah. Obviously shared code can go into a utility function if there is enough to justify one.

Mr0grog replied:
@danielballan how’s this? https://github.com/edgi-govdata-archiving/web-monitoring-processing/pull/153/commits/33a1b0ead5a0a616027b0128b73967f0dc521142

Am I bending over too far backwards to accommodate a formatting that is only theoretical?

I made the routes `/links` and `/links_json` (instead of `/links.json`) because I realized they are both JSON HTTP responses at the end of the day, so putting an extension on one implied the other was not JSON, even though it is, and blah blah blah confusing.

danielballan replied:
I think there's a good chance we'll be glad we have this.

Mr0grog commented:
Oh definitely, I don’t think this is good to merge because it _absolutely_ needs design work to address the color issues. Was just getting the obvious stuff that doesn’t require much thinking out of the way first :)

Mr0grog commented:
Summary from meeting last night (sorry I didn’t get this written up sooner):

I haven’t addressed color yet because I think that’s a pervasive issue across all diffs. However! There are lots of other non-color cues we can use in this diff that I’ve experimented with adding. The main changes I’ve made are:

1. Using differing bullet points for different changes:
    - “⚬” (an empty circle) for no change
    - “+” when a listing is completely new
    - “-” when a listing is completely removed
    - “±” when a listing has *changed* (either the text or the URL was modified)

2. Split listings for *changed* links into two rows, the first listing the removals and the second listing the additions. That makes the actual old vs. new content easy to discern. The text and URL are each stacked on top of each other, so they are also easy to compare. We still highlight *what* was added and removed within the text or URL.

So that basically looks like:

![screen shot 2018-02-06 at 10 35 37 am](https://user-images.githubusercontent.com/74178/35878577-95ce03f8-0b2d-11e8-9222-bf5215485a77.png)

Then we’ve got some slight tweaks—I’m curious how people feel about these variations:

- Vertically center (instead of top-aligning) the text/URL that hasn’t changed.

    ![screen shot 2018-02-06 at 10 39 16 am](https://user-images.githubusercontent.com/74178/35878649-c29e8f92-0b2d-11e8-85ce-a44201be91c0.png)

- Completely separate text and URL into a more table-like view, where each is in a separate column.

    ![screen shot 2018-02-06 at 10 58 04 am](https://user-images.githubusercontent.com/74178/35878689-df7d2bd2-0b2d-11e8-9d71-9eaff9da7f7a.png)

Mr0grog commented:
Moved the color issue (because I think it affects things more widely) to #158

lightandluck commented:
I like the center alignment and table-view. Feels like it makes it easier to "scan". 😉  The symbology is easy to understand. Overall, looking awesome and improves this useful feature!

Mr0grog commented:
Probably going to wait to finish this until #155 is in because they overlap a bit and will conflict.

Mr0grog commented:
OK, I think this is set to go. Since this has been sitting around so long, I think I’m going to go ahead and merge it in a day or so if nobody has any feedback. Overall, this should represent some pretty serious improvements in this diff, both on accuracy and readability.

Current output now looks like:

<img width="978" alt="screen shot 2018-06-19 at 12 03 47 am" src="https://user-images.githubusercontent.com/74178/41583130-6b566122-7358-11e8-9513-9bcc72fb9995.png">

NOTE: the test failures are Internet Archive. I can’t tell if they’re just having intermittent issues right now or if they might be blocking CircleCI or AWS. The same tests run fine for me locally. In any case, the changes here should not affect any of that code at all.

Mr0grog commented:
I’m going ahead and merging this. We can do cleanup in subsequent PRs. I don’t want it sitting pointlessly forever again.
